shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/model/Dataset.java

```diff
@@ -165,7 +149,14 @@ public abstract class Dataset extends AbstractEntity {
 	@JoinTable(name = "DATASET_TAG", joinColumns = @JoinColumn(name = "DATASET_ID"), inverseJoinColumns = @JoinColumn(name = "STUDY_TAG_ID"))
 	private List<StudyTag> tags;
 
-	private Long sourceId;
+	@ManyToOne(fetch = FetchType.LAZY)
+	@JoinColumn(name = "source_id")
+	private Dataset source;
+
+	@OneToMany(fetch = FetchType.LAZY, mappedBy = "source", cascade = CascadeType.ALL)
+	private List<Dataset> copies;
+
+    private Integer origin;
```

Comment:
Several problems with origin imo : 
* I'm not sure it's needed here as a Dataset property. This information can be deduced by the state of ```Dataset source``` and ```List<Dataset> copies```. If it has copies it's a sources, if it has a source it's a copy. If the getter ```getOrigin()``` is usefull, it can infer its return using source and copies.
* An entity could be both a copy and a source. So you could either use a 3rd state in the enum, or have two booleans instead : isCopy and isSource.
* If origin is removed from Dataset and EntityOrigin is not to be persisted in the database, it does not need to have an id, a getType method or a constructor.

Reply:
- Indeed, EntityOrigin is not needed since we can check the content of copies and source. I'll remove this enum and work with copies/source only.

- A copy can't be copied.